src/lib/gl_engine/tvgGlTessellator.cpp

```diff
@@ -0,0 +1,2150 @@
+    LinkedList<Edge>::Remove<&Edge::below_prev, &Edge::below_next>(this, &top->edge_below.head, &top->edge_below.tail);
+    // update top vertex
+    top = v;
+    // recompute line equation
+    recompute();
+    // insert self to new top's below list
+    top->insertBelow(this);
+}
+
+static void remove_edge_above(Edge *edge)
+{
+    LinkedList<Edge>::Remove<&Edge::above_prev, &Edge::above_next>(edge, &edge->bottom->edge_above.head,
+                                                                   &edge->bottom->edge_above.tail);
+}
+
+static void remove_edge_below(Edge *edge)
+{
+    LinkedList<Edge>::Remove<&Edge::below_prev, &Edge::below_next>(edge, &edge->top->edge_below.head,
+                                                                   &edge->top->edge_below.tail);
+}
+
+void Edge::disconnect()
+{
+    remove_edge_above(this);
+    remove_edge_below(this);
+}
+
+void ActiveEdgeList::insert(Edge *e, Edge *prev, Edge *next)
+{
+    LinkedList<Edge>::Insert<&Edge::left, &Edge::right>(e, prev, next, &head, &tail);
+    if (!valid()) {
+        return;
+    }
+}
+
+void ActiveEdgeList::insert(Edge *e, Edge *prev)
+{
+    auto next = prev ? prev->right : head;
+
+    insert(e, prev, next);
+}
+
+void ActiveEdgeList::append(Edge *e)
+{
+    insert(e, tail, nullptr);
+}
+
+void ActiveEdgeList::remove(Edge *e)
+{
+    LinkedList<Edge>::Remove<&Edge::left, &Edge::right>(e, &head, &tail);
+}
+
+bool ActiveEdgeList::contains(Edge *edge)
+{
+    return edge->left || edge->right || head == edge;
+}
+
+void ActiveEdgeList::rewind(Vertex **current, Vertex *dst)
+{
+    if (!current || *current == dst || VertexCompare::Compare((*current)->point, dst->point)) {
+        return;
+    }
+
+    Vertex *v = *current;
+
+    while (v != dst) {
+        v = v->prev;
+
+        for (auto e = v->edge_below.head; e; e = e->below_next) {
+            this->remove(e);
+        }
+
+        auto left = v->left;
+
+        for (auto e = v->edge_above.head; e; e = e->above_next) {
+            this->insert(e, left);
+            left = e;
+
+            auto top = e->top;
+            if (VertexCompare::Compare(top->point, dst->point) &&
+                ((top->left && !top->left->isLeftOf(e->top->point)) ||
+                 (top->right && !top->right->isRightOf(e->top->point)))) {
+                dst = top;
+            }
+        }
+    }
+
+    *current = v;
+}
+
+void ActiveEdgeList::findEnclosing(Vertex *v, Edge **left, Edge **right)
+{
+    if (v->edge_above.head && v->edge_above.tail) {
+        *left = v->edge_above.head->left;
+        *right = v->edge_above.tail->right;
+        return;
+    }
+
+    Edge *prev = nullptr;
+    Edge *next = nullptr;
+
+    // walk through aet to get left most edge
+    for (prev = tail; prev != nullptr; prev = prev->left) {
+        if (prev->isLeftOf(v->point)) {
+            break;
+        }
+        next = prev;
+    }
+
+    *left = prev;
+    *right = next;
+}
+
+bool _validEdgePair(Edge* left, Edge* right) {
```

Comment:
Please add the 'static' keyword to any of the internal methods.
bool _validEdgePair() -> static bool _validEdgePair()